# Understanding EV Adoption
#A Data-Driven Analysis of Range Anxiety and Consumer Purchase Decisions

## Data Profiling
### Objectivey

#The objective of this notebook is to understand the structure, quality, and completeness of the consumer survey dataset before performing any data cleaning or analysis.


In [1]:
## Import Libraries
import pandas as pd
import numpy as np

In [2]:
## Load Dataset

df = pd.read_excel("../data/raw/EV_Adoption_and_Range_Anxiety_Dataset.xlsx")

In [3]:
## Data Types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Buyer_ID                     10000 non-null  str    
 1   Age                          10000 non-null  int64  
 2   Gender                       10000 non-null  str    
 3   Annual_Income_USD            9822 non-null   float64
 4   City_Type                    10000 non-null  str    
 5   Daily_Commute_km             9819 non-null   float64
 6   Number_of_Cars_Owned         10000 non-null  int64  
 7   Current_Car_Type             10000 non-null  str    
 8   Charging_Stations_Near_Home  10000 non-null  int64  
 9   Charging_Stations_Near_Work  10000 non-null  int64  
 10  Home_Charging_Possible       10000 non-null  str    
 11  Environmental_Concern_Level  9816 non-null   float64
 12  Subsidy_Available            10000 non-null  str    
 13  Range_Anxiety_Level         

In [4]:
import os

print(os.listdir(".."))


['.git', '.gitignore', 'Business_Questions.md', 'data', 'eda', 'EV_Adoption_and_Range_Anxiety_Dataset.xlsx', 'Project_Plan.md', 'README.md', 'reports', 'SQL']


In [5]:
import os

print(os.listdir("../data/raw"))

['EV_Adoption_and_Range_Anxiety_Dataset.xlsx']


In [6]:
import pandas as pd

df = pd.read_excel("../data/raw/EV_Adoption_and_Range_Anxiety_Dataset.xlsx")

df.head()

,Buyer_ID,Age,Gender,Annual_Income_USD,City_Type,Daily_Commute_km,Number_of_Cars_Owned,Current_Car_Type,Charging_Stations_Near_Home,Charging_Stations_Near_Work,Home_Charging_Possible,Environmental_Concern_Level,Subsidy_Available,Range_Anxiety_Level,Will_Buy_EV
0,EV00001,56,Male,44906.0,Rural,18.1,1,Truck,1,0,Yes,4.0,Yes,Low,No
1,EV00002,36,Male,88754.0,Rural,5.0,1,SUV,2,3,Yes,1.0,Yes,Low,No
2,EV00003,42,Male,155958.0,Urban,32.2,2,Sedan,14,3,No,2.0,Yes,Medium,Yes
3,EV00004,31,Male,37059.0,Rural,89.9,1,Sedan,2,1,Yes,4.0,No,Low,No
4,EV00005,48,Female,48552.0,Rural,71.0,1,Sedan,2,0,Yes,3.0,Yes,Low,No


In [7]:
df.shape

(10000, 15)

In [8]:
df.columns

Index(['Buyer_ID', 'Age', 'Gender', 'Annual_Income_USD', 'City_Type',
       'Daily_Commute_km', 'Number_of_Cars_Owned', 'Current_Car_Type',
       'Charging_Stations_Near_Home', 'Charging_Stations_Near_Work',
       'Home_Charging_Possible', 'Environmental_Concern_Level',
       'Subsidy_Available', 'Range_Anxiety_Level', 'Will_Buy_EV'],
      dtype='str')

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Buyer_ID                     10000 non-null  str    
 1   Age                          10000 non-null  int64  
 2   Gender                       10000 non-null  str    
 3   Annual_Income_USD            9822 non-null   float64
 4   City_Type                    10000 non-null  str    
 5   Daily_Commute_km             9819 non-null   float64
 6   Number_of_Cars_Owned         10000 non-null  int64  
 7   Current_Car_Type             10000 non-null  str    
 8   Charging_Stations_Near_Home  10000 non-null  int64  
 9   Charging_Stations_Near_Work  10000 non-null  int64  
 10  Home_Charging_Possible       10000 non-null  str    
 11  Environmental_Concern_Level  9816 non-null   float64
 12  Subsidy_Available            10000 non-null  str    
 13  Range_Anxiety_Level         

In [10]:
df.describe()

,Age,Annual_Income_USD,Daily_Commute_km,Number_of_Cars_Owned,Charging_Stations_Near_Home,Charging_Stations_Near_Work,Environmental_Concern_Level
count,10000.000000,9822.000000,9819.000000,10000.000000,10000.000000,10000.000000,9816.000000
mean,46.936200,85378.485848,41.105418,1.859900,5.346900,7.458200,2.982783
std,12.945968,32838.684167,23.351276,0.855421,4.049074,5.259374,1.413784
min,25.000000,30000.000000,5.000000,1.000000,0.000000,0.000000,1.000000
25%,36.000000,61369.500000,23.250000,1.000000,2.000000,3.000000,2.000000
50%,47.000000,84708.000000,40.200000,2.000000,5.000000,6.000000,3.000000
75%,58.000000,107383.250000,57.000000,2.000000,8.000000,11.000000,4.000000
max,69.000000,223345.000000,135.500000,4.000000,14.000000,19.000000,5.000000


In [11]:
df.isnull().sum()


Buyer_ID                         0
Age                              0
Gender                           0
Annual_Income_USD              178
City_Type                        0
Daily_Commute_km               181
Number_of_Cars_Owned             0
Current_Car_Type                 0
Charging_Stations_Near_Home      0
Charging_Stations_Near_Work      0
Home_Charging_Possible           0
Environmental_Concern_Level    184
Subsidy_Available                0
Range_Anxiety_Level              0
Will_Buy_EV                      0
dtype: int64

## Initial Data Quality Assessment

#Observations:

#- Dataset contains  178+181+184 missing values.
#- Missing values are present in 3 columns.
#- No action has been taken yet because profiling should precede cleaning.

In [12]:
df["Will_Buy_EV"].value_counts()

Will_Buy_EV
No     8250
Yes    1750
Name: count, dtype: int64

In [13]:
df["Range_Anxiety_Level"].value_counts()


Range_Anxiety_Level
Low       8310
Medium    1523
High       167
Name: count, dtype: int64

In [14]:
pd.crosstab(df["Range_Anxiety_Level"], df["Will_Buy_EV"])
#relation between two columns

Will_Buy_EV,No,Yes
Range_Anxiety_Level,,
High,166,1
Low,6659,1651
Medium,1425,98


In [15]:
pd.crosstab(
    df["Range_Anxiety_Level"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Range_Anxiety_Level,,
High,99.401198,0.598802
Low,80.132371,19.867629
Medium,93.565332,6.434668


In [16]:
pd.crosstab(
    df["Home_Charging_Possible"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Home_Charging_Possible,,
No,86.630180,13.369820
Yes,79.928606,20.071394


In [17]:
pd.crosstab(
    df["Environmental_Concern_Level"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Environmental_Concern_Level,,
1.0,97.863885,2.136115
2.0,95.071869,4.928131
3.0,85.546476,14.453524
4.0,74.710910,25.289090
5.0,58.595388,41.404612


In [18]:
df["Annual_Income_USD"].describe()

count      9822.000000
mean      85378.485848
std       32838.684167
min       30000.000000
25%       61369.500000
50%       84708.000000
75%      107383.250000
max      223345.000000
Name: Annual_Income_USD, dtype: float64

In [19]:
df["Annual_Income_USD"].isnull().sum()

np.int64(178)

In [20]:
df["Income_Group"] = pd.qcut(
    df["Annual_Income_USD"],
    q=4,
    labels=["Low", "Middle", "Upper Middle", "High"]
)

In [21]:
pd.crosstab(
    df["Income_Group"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Income_Group,,
Low,90.472313,9.527687
Middle,85.824847,14.175153
Upper Middle,80.488798,19.511202
High,73.778502,26.221498


In [22]:
bins = [20, 30, 40, 50, 60, 70]
labels = ["20-29", "30-39", "40-49", "50-59", "60-69"]

df["Age_Group"] = pd.cut(df["Age"], bins=bins, labels=labels, right=False)

In [23]:
pd.crosstab(
    df["Age_Group"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Age_Group,,
20-29,82.479784,17.520216
30-39,82.935615,17.064385
40-49,83.018038,16.981962
50-59,80.527753,19.472247
60-69,83.507973,16.492027


In [24]:
pd.crosstab(
    df["Subsidy_Available"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Subsidy_Available,,
No,97.469940,2.530060
Yes,72.553262,27.446738


In [25]:
pd.crosstab(
    df["Current_Car_Type"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Current_Car_Type,,
Hatchback,82.534471,17.465529
SUV,81.899025,18.100975
Sedan,82.775000,17.225000
Truck,83.451060,16.548940


### Observation

EV purchase intention varies only slightly across current car types. SUV owners show the highest purchase intention (18.10%), while truck owners show the lowest (16.55%). The relatively small difference suggests that current car type may not be a strong differentiating factor in EV purchase intention.


In [26]:
bins = [0, 10, 20, 30, 40, 50, float("inf")]
labels = ["0-9", "10-19", "20-29", "30-39", "40-49", "50+"]

df["Commute_Group"] = pd.cut(
    df["Daily_Commute_km"],
    bins=bins,
    labels=labels,
    right=False
)

pd.crosstab(
    df["Commute_Group"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Commute_Group,,
0-9,80.965392,19.034608
10-19,80.487805,19.512195
20-29,82.816048,17.183952
30-39,82.298547,17.701453
40-49,82.364597,17.635403
50+,83.639545,16.360455


### Observation

EV purchase intention is highest among respondents with daily commutes of 10–19 km (19.51%) and lowest among those commuting 50+ km (16.36%). Purchase intention does not increase with commute distance, suggesting that longer daily commutes alone may not be a strong driver of EV purchase intention in this dataset.

In [27]:
pd.crosstab(
    df["Charging_Stations_Near_Home"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Charging_Stations_Near_Home,,
0,82.188841,17.811159
1,80.227743,19.772257
2,81.259601,18.740399
3,84.500574,15.499426
4,82.976040,17.023960
5,84.568651,15.431349
6,84.082625,15.917375
7,84.670659,15.329341
8,80.428954,19.571046


In [28]:
bins = [-1, 4, 9, 14]
labels = ["Low", "Medium", "High"]

df["Home_Charging_Group"] = pd.cut(
    df["Charging_Stations_Near_Home"],
    bins=bins,
    labels=labels
)

In [29]:
pd.crosstab(
    df["Home_Charging_Group"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Home_Charging_Group,,
Low,82.092928,17.907072
Medium,83.472351,16.527649
High,81.885203,18.114797


### Observation

EV purchase intention varies only moderately across home charging availability groups. The High group has the highest purchase intention (18.11%), followed closely by the Low group (17.91%), while the Medium group has the lowest (16.53%). This suggests that the number of charging stations near home alone does not show a clear linear relationship with EV purchase intention in this dataset.

In [30]:
df["Charging_Stations_Near_Work"].min(), df["Charging_Stations_Near_Work"].max()

(np.int64(0), np.int64(19))

In [31]:
bins = [-1, 6, 13, 19]
labels = ["Low", "Medium", "High"]

df["Work_Charging_Group"] = pd.cut(
    df["Charging_Stations_Near_Work"],
    bins=bins,
    labels=labels
)

In [32]:
pd.crosstab(
    df["Work_Charging_Group"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Work_Charging_Group,,
Low,82.481023,17.518977
Medium,82.676444,17.323556
High,82.219640,17.780360


### Observation

EV purchase intention is very similar across workplace charging availability groups, ranging from 17.32% to 17.78%. This suggests that the number of charging stations near work does not show a meaningful relationship with EV purchase intention in this dataset.

In [33]:
df["Number_of_Cars_Owned"].min(), df["Number_of_Cars_Owned"].max()

(np.int64(1), np.int64(4))

In [34]:
pd.crosstab(
    df["Number_of_Cars_Owned"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Number_of_Cars_Owned,,
1,82.742017,17.257983
2,82.388060,17.611940
3,81.286934,18.713066
4,85.127202,14.872798


### Observation

EV purchase intention varies across the number of cars owned. Respondents owning three cars show the highest purchase intention (18.71%), while those owning four cars show the lowest (14.87%). The relationship is not consistently increasing with the number of cars owned, suggesting that car ownership alone may not be a strong linear driver of EV purchase intention.

In [35]:
pd.crosstab(
    df["Gender"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
Gender,,
Female,81.969764,18.030236
Male,82.907133,17.092867
Other,83.388704,16.611296


### Observation

EV purchase intention varies only slightly across gender groups. Female respondents have the highest purchase intention (18.03%), followed by male respondents (17.09%) and other respondents (16.61%). The relatively small differences suggest that gender alone may not be a strong differentiating factor in EV purchase intention.

In [36]:
pd.crosstab(
    df["City_Type"],
    df["Will_Buy_EV"],
    normalize="index"
) * 100

Will_Buy_EV,No,Yes
City_Type,,
Rural,81.016043,18.983957
Suburban,82.318271,17.681729
Urban,83.080348,16.919652


### Observation

EV purchase intention varies across city types. Rural respondents have the highest purchase intention (18.98%), followed by suburban respondents (17.68%), while urban respondents have the lowest (16.92%). Although the difference is noticeable, city type alone should not be interpreted as a causal driver of EV purchase intention.

In [37]:
df["Will_Buy_EV"].value_counts(normalize=True) * 100

Will_Buy_EV
No     82.5
Yes    17.5
Name: proportion, dtype: float64

### Observation

Range anxiety shows a strong relationship with EV purchase intention. Respondents with Low range anxiety have the highest purchase intention (19.87%), compared with 6.43% among those with Medium range anxiety and only 0.60% among those with High range anxiety. This suggests that reducing consumers' concerns about driving range could be an important consideration when encouraging EV adoption.

### Observation

Environmental concern shows a strong positive relationship with EV purchase intention. Purchase intention increases consistently from 2.14% among respondents with the lowest environmental concern (Level 1) to 41.40% among respondents with the highest environmental concern (Level 5). This suggests that environmentally conscious consumers represent a potentially important segment for EV adoption.

# EDA Summary

## Key Findings

### 1. Environmental Concern
Environmental concern shows a strong positive relationship with EV purchase intention. Purchase intention increases from 2.14% at Level 1 to 41.40% at Level 5.

### 2. Range Anxiety
Range anxiety shows a strong negative relationship with EV purchase intention. Purchase intention decreases from 19.87% for Low range anxiety to only 0.60% for High range anxiety.

### 3. Subsidy Availability
Respondents with a subsidy available have substantially higher EV purchase intention (27.45%) compared with respondents without a subsidy (2.53%).

### 4. Income
EV purchase intention increases across income groups, from 9.53% among Low-income respondents to 26.22% among High-income respondents.

### 5. Other Variables
Gender, current car type, city type, number of cars owned, commute distance, home charging availability, and workplace charging availability show relatively smaller differences in EV purchase intention.

## Overall Conclusion

The strongest patterns in the descriptive analysis are associated with environmental concern, range anxiety, subsidy availability, and income. These variables should receive greater attention in subsequent analysis and business recommendations.

In [38]:
df.isnull().sum()

Buyer_ID                         0
Age                              0
Gender                           0
Annual_Income_USD              178
City_Type                        0
Daily_Commute_km               181
Number_of_Cars_Owned             0
Current_Car_Type                 0
Charging_Stations_Near_Home      0
Charging_Stations_Near_Work      0
Home_Charging_Possible           0
Environmental_Concern_Level    184
Subsidy_Available                0
Range_Anxiety_Level              0
Will_Buy_EV                      0
Income_Group                   178
Age_Group                        0
Commute_Group                  181
Home_Charging_Group              0
Work_Charging_Group              0
dtype: int64

In [39]:
#check the percentage of missing values
(df.isnull().sum() / len(df) * 100).round(2)

Buyer_ID                       0.00
Age                            0.00
Gender                         0.00
Annual_Income_USD              1.78
City_Type                      0.00
Daily_Commute_km               1.81
Number_of_Cars_Owned           0.00
Current_Car_Type               0.00
Charging_Stations_Near_Home    0.00
Charging_Stations_Near_Work    0.00
Home_Charging_Possible         0.00
Environmental_Concern_Level    1.84
Subsidy_Available              0.00
Range_Anxiety_Level            0.00
Will_Buy_EV                    0.00
Income_Group                   1.78
Age_Group                      0.00
Commute_Group                  1.81
Home_Charging_Group            0.00
Work_Charging_Group            0.00
dtype: float64

In [40]:
df[[
    "Annual_Income_USD",
    "Daily_Commute_km",
    "Environmental_Concern_Level"
]].describe()

,Annual_Income_USD,Daily_Commute_km,Environmental_Concern_Level
count,9822.000000,9819.000000,9816.000000
mean,85378.485848,41.105418,2.982783
std,32838.684167,23.351276,1.413784
min,30000.000000,5.000000,1.000000
25%,61369.500000,23.250000,2.000000
50%,84708.000000,40.200000,3.000000
75%,107383.250000,57.000000,4.000000
max,223345.000000,135.500000,5.000000


In [41]:
df_clean = df.copy()

In [42]:
df_clean = df.copy()


In [43]:
df_clean.shape

(10000, 20)

In [44]:
df_clean.isnull().sum()

Buyer_ID                         0
Age                              0
Gender                           0
Annual_Income_USD              178
City_Type                        0
Daily_Commute_km               181
Number_of_Cars_Owned             0
Current_Car_Type                 0
Charging_Stations_Near_Home      0
Charging_Stations_Near_Work      0
Home_Charging_Possible           0
Environmental_Concern_Level    184
Subsidy_Available                0
Range_Anxiety_Level              0
Will_Buy_EV                      0
Income_Group                   178
Age_Group                        0
Commute_Group                  181
Home_Charging_Group              0
Work_Charging_Group              0
dtype: int64

In [45]:
#filling the missing value with the median values

df_clean["Annual_Income_USD"] = df_clean["Annual_Income_USD"].fillna(
    df_clean["Annual_Income_USD"].median()
)

In [46]:
df_clean["Annual_Income_USD"].isnull().sum()

np.int64(0)

In [47]:
df_clean["Daily_Commute_km"] = df_clean["Daily_Commute_km"].fillna(
    df_clean["Daily_Commute_km"].median()
)

In [48]:
df_clean["Daily_Commute_km"].isnull().sum()

np.int64(0)

In [49]:
df_clean["Environmental_Concern_Level"] = df_clean["Environmental_Concern_Level"].fillna(
    df_clean["Environmental_Concern_Level"].median()
)

In [50]:
df_clean["Environmental_Concern_Level"].isnull().sum()

np.int64(0)

In [51]:
df_clean.isnull().sum()

Buyer_ID                         0
Age                              0
Gender                           0
Annual_Income_USD                0
City_Type                        0
Daily_Commute_km                 0
Number_of_Cars_Owned             0
Current_Car_Type                 0
Charging_Stations_Near_Home      0
Charging_Stations_Near_Work      0
Home_Charging_Possible           0
Environmental_Concern_Level      0
Subsidy_Available                0
Range_Anxiety_Level              0
Will_Buy_EV                      0
Income_Group                   178
Age_Group                        0
Commute_Group                  181
Home_Charging_Group              0
Work_Charging_Group              0
dtype: int64

In [52]:
#income group
df_clean["Income_Group"] = pd.qcut(
    df_clean["Annual_Income_USD"],
    q=4,
    labels=["Low", "Middle", "Upper Middle", "High"]
)

In [53]:
#commute group
bins = [0, 10, 20, 30, 40, 50, float("inf")]
labels = ["0-9", "10-19", "20-29", "30-39", "40-49", "50+"]

df_clean["Commute_Group"] = pd.cut(
    df_clean["Daily_Commute_km"],
    bins=bins,
    labels=labels,
    right=False
)

In [54]:
df_clean.isnull().sum()

Buyer_ID                       0
Age                            0
Gender                         0
Annual_Income_USD              0
City_Type                      0
Daily_Commute_km               0
Number_of_Cars_Owned           0
Current_Car_Type               0
Charging_Stations_Near_Home    0
Charging_Stations_Near_Work    0
Home_Charging_Possible         0
Environmental_Concern_Level    0
Subsidy_Available              0
Range_Anxiety_Level            0
Will_Buy_EV                    0
Income_Group                   0
Age_Group                      0
Commute_Group                  0
Home_Charging_Group            0
Work_Charging_Group            0
dtype: int64

In [55]:
df_clean.duplicated().sum()

np.int64(0)

In [56]:
#checking if buyer id is unique
df_clean["Buyer_ID"].nunique()

10000

In [57]:
len(df_clean)

10000

In [58]:
df_clean["Age"].min(), df_clean["Age"].max()

(np.int64(25), np.int64(69))

### Age Validation

Age values range from 25 to 69 years, with no obviously invalid or impossible values identified.

In [59]:
df_clean["Annual_Income_USD"].min(), df_clean["Annual_Income_USD"].max()

(np.float64(30000.0), np.float64(223345.0))

### Annual Income Validation

Annual income ranges from $30,000 to $223,345. No negative or obviously invalid income values were identified.

In [60]:
df_clean["Daily_Commute_km"].min(), df_clean["Daily_Commute_km"].max()

(np.float64(5.0), np.float64(135.5))

### Daily Commute Validation

Daily commute distance ranges from 5 km to 135.5 km. No negative or zero commute distances were identified.

In [61]:
df_clean["Environmental_Concern_Level"].min(), df_clean["Environmental_Concern_Level"].max()

(np.float64(1.0), np.float64(5.0))

### Environmental Concern Validation

Environmental concern scores range from 1 to 5, which matches the expected scale. No values outside the valid range were identified.

In [62]:
df_clean["Will_Buy_EV"].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

### Will_Buy_EV Validation

The target variable contains two valid categories: `Yes` and `No`. No unexpected or inconsistent category values were identified.

In [63]:
df_clean["Gender"].unique()

<StringArray>
['Male', 'Female', 'Other']
Length: 3, dtype: str

### Gender Validation

The Gender variable contains three categories: `Male`, `Female`, and `Others`. No unexpected categories were identified.

In [64]:
df_clean["City_Type"].unique()

<StringArray>
['Rural', 'Urban', 'Suburban']
Length: 3, dtype: str

### City Type Validation

The City Type variable contains three categories: `Rural`, `Urban`, and `Suburban`. No unexpected categories were identified.

In [65]:
df_clean["Current_Car_Type"].unique()

<StringArray>
['Truck', 'SUV', 'Sedan', 'Hatchback']
Length: 4, dtype: str

### Current Car Type Validation

The Current Car Type variable contains four categories: `Truck`, `SUV`, `Sedan`, and `Hatchback`. No unexpected categories were identified.

In [66]:
df_clean["Range_Anxiety_Level"].unique()

<StringArray>
['Low', 'Medium', 'High']
Length: 3, dtype: str

### Range Anxiety Validation

The Range Anxiety Level variable contains three categories: `Low`, `Medium`, and `High`. No unexpected categories were identified.


In [67]:
df_clean["Income_Group"].unique()


['Low', 'Upper Middle', 'High', 'Middle']
Categories (4, str): ['Low' < 'Middle' < 'Upper Middle' < 'High']

### Income Group Validation

The Income Group variable contains four ordered categories: `Low`, `Middle`, `Upper Middle`, and `High`. The categories follow the expected income progression from Low to High, with no unexpected categories identified.

In [68]:
df_clean["Age_Group"].unique()

['50-59', '30-39', '40-49', '60-69', '20-29']
Categories (5, str): ['20-29' < '30-39' < '40-49' < '50-59' < '60-69']

### Age Group Validation

The Age Group variable contains five ordered categories: `20-29`, `30-39`, `40-49`, `50-59`, and `60-69`. The categories follow the expected chronological order, with no unexpected categories identified.

In [69]:
df_clean["Home_Charging_Group"].unique()

['Low', 'High', 'Medium']
Categories (3, str): ['Low' < 'Medium' < 'High']

### Home Charging Group Validation

The Home Charging Group variable contains three categories: `Low`, `Medium`, and `High`. No unexpected categories were identified.

In [70]:
df_clean["Work_Charging_Group"].unique()

['Low', 'Medium', 'High']
Categories (3, str): ['Low' < 'Medium' < 'High']

In [71]:
df_clean.shape

(10000, 20)

In [72]:
df_clean.to_csv("../data/cleaned/EV_Adoption_Cleaned.csv", index=False)

In [73]:
import os

os.makedirs("../data/cleaned", exist_ok=True)

In [74]:
df_clean.to_csv("../data/cleaned/EV_Adoption_Cleaned.csv", index=False)

In [75]:
os.path.exists("../data/cleaned/EV_Adoption_Cleaned.csv")

True

In [76]:
pd.read_csv("../data/cleaned/EV_Adoption_Cleaned.csv").shape

(10000, 20)